storage/rocksdb/ha_rocksdb.cc

```diff
@@ -9556,6 +9809,247 @@ int ha_rocksdb::index_read_intern(uchar *const buf, const uchar *const key,
+    else if (find_flag == HA_READ_AFTER_KEY)
+      start_has_inf_suffix = true;
+    else {
+      // Unknown type of range, shouldn't happen
+      DBUG_ASSERT(0);
+      big_range = true;
+    }
+
+    // Known end range bounds: HA_READ_AFTER_KEY, HA_READ_BEFORE_KEY
+    if (end_key->flag == HA_READ_AFTER_KEY) {
+      // this is "key_part <= const".
+      end_has_inf_suffix = true;
+    } else if (end_key->flag == HA_READ_BEFORE_KEY) {
+      // this is "key_part < const", non-inclusive.
+      end_has_inf_suffix = false;
+    } else {
+      // Unknown type of range, shouldn't happen
+      DBUG_ASSERT(0);
+      big_range = true;
+    }
+
+    uchar pack_buffer[MAX_KEY_LENGTH];
+    uint end_slice_size = kd.pack_index_tuple(
+        table, pack_buffer, end_slice_buf, end_key->key, end_key->keypart_map);
+
+    end_slice =
+        rocksdb::Slice(reinterpret_cast<char *>(end_slice_buf), end_slice_size);
+  } else {
+    big_range = true;
+#if 0
+    // The below is code to handle this without LockingIterator:
+    // No end key
+    // Known start range bounds: HA_READ_KEY_OR_NEXT, HA_READ_AFTER_KEY
+    if (find_flag == HA_READ_KEY_OR_NEXT)
+      start_has_inf_suffix= false;
+    else if (find_flag == HA_READ_AFTER_KEY)
+      start_has_inf_suffix= true;
+    else
+      DBUG_ASSERT(0);
+
+    uint end_slice_size;
+    kd.get_infimum_key(end_slice_buf, &end_slice_size);
+    end_slice= rocksdb::Slice((char*)end_slice_buf, end_slice_size);
+    end_has_inf_suffix= true;
+#endif
+  }
+
+  if (big_range) {
+    *use_locking_iterator = true;
+    return 0;
+  }
+
+  rocksdb::Endpoint start_endp;
+  rocksdb::Endpoint end_endp;
+
+  if (kd.m_is_reverse_cf) {
+    // Flip the endpoint flag values, as explained in the
+    // RangeFlagsShouldBeFlippedForRevCF comment above.
+    start_endp = rocksdb::Endpoint(end_slice, !end_has_inf_suffix);
+    end_endp = rocksdb::Endpoint(slice, !start_has_inf_suffix);
+  } else {
+    start_endp = rocksdb::Endpoint(slice, start_has_inf_suffix);
+    end_endp = rocksdb::Endpoint(end_slice, end_has_inf_suffix);
+  }
+
+  /*
+    RocksDB's iterator is reading the snapshot of the data that was taken at
+    the time the iterator was created.
+
+    After we've got a lock on the range, we'll need to refresh the iterator
+    to read the latest contents. (If we use the iterator created before the
+    lock_range() call, we may miss the changes that were made/committed after
+    the iterator was created but before the lock_range() call was made).
+
+    RocksDB has Iterator::Refresh() method, but alas, it is not implemented for
+    the iterator returned by Transaction object (Transaction object returns
+    BaseDeltaIterator which allows one to see the transactions's own changes).
+
+    Our solution to this is to release the iterator and create the new one.
+    We release it here, it will be created as soon as there's a need to read
+    records.
+  */
+  m_iterator->reset();
```

Comment:
One option that we can consider is to use the tailing option in rocksdb iterators. It may be worth capturing this in a comment so that we don't lose this idea.

Actually, I thinking tailing won't work because it will only read from files from its superversion. I guess we do just need iterator refresh after all.